cpp/src/groupby/sort/group_topk.cu

```diff
@@ -0,0 +1,111 @@
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/sizes_to_offsets_iterator.cuh>
+#include <cudf/detail/sorting.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/types.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+
+#include <cuda/functional>
+#include <thrust/transform_scan.h>
+
+#include <memory>
+
+namespace cudf {
+namespace groupby {
+namespace detail {
+
+__global__ void compute_topk_indices(cudf::device_span<size_type const> group_sizes,
+                                     cudf::device_span<size_type const> group_offsets,
+                                     cudf::device_span<size_type const> index_offsets,
+                                     cudf::device_span<size_type> indices,
+                                     size_type num_groups,
+                                     size_type k)
+{
+  // warp per group writes output indices
+  for (thread_index_type idx = threadIdx.x + blockDim.x * blockIdx.x;
+       idx < cudf::detail::warp_size * num_groups;
+       idx += gridDim.x * blockDim.x) {
+    auto const group{idx / cudf::detail::warp_size};
+    auto const lane{idx % cudf::detail::warp_size};
+    if (group >= num_groups) break;
+    auto const group_size   = group_sizes[group];
+    auto const k_loc        = min(group_size, k);
+    auto const index_offset = group_offsets[group];
+    auto const out_offset   = index_offsets[group];
+    for (size_type off = lane; off < k_loc; off += cudf::detail::warp_size) {
+      indices[out_offset + off] = index_offset + off;
+    }
+  }
+}
+
+std::unique_ptr<column> group_topk(column_view const& values,
+                                   cudf::device_span<size_type const> group_sizes,
+                                   cudf::device_span<size_type const> group_offsets,
+                                   size_type num_groups,
+                                   size_type k,
+                                   order order,
+                                   rmm::cuda_stream_view stream,
+                                   rmm::mr::device_memory_resource* mr)
+{
+  auto group_size_fn = cuda::proclaim_return_type<size_type>(
+    [group_sizes, num_groups, k] __device__(size_type i) -> size_type {
+      return i < num_groups ? min(k, group_sizes[i]) : size_type{0};
+    });
+  auto size_per_group = cudf::detail::make_counting_transform_iterator(0, group_size_fn);
+  auto [offsets_column, output_size] = cudf::detail::make_offsets_child_column(
+    size_per_group, size_per_group + num_groups, stream, mr);
+  auto indices = rmm::device_uvector<size_type>(output_size, stream);
+  cudf::detail::grid_1d const config{num_groups, 128};
+  compute_topk_indices<<<config.num_blocks, config.num_threads_per_block, 0, stream.value()>>>(
+    group_sizes, group_offsets, offsets_column->view(), indices, num_groups, k);
```

Comment:
I couldn't think hard enough to see how to write this as a fancy thrust scan, so wrote a kernel by hand. But would love for a better solution here.